### PR TITLE
[SD-1422] - Migrate Spree Digital into Spree Core

### PR DESCRIPTION
**Pull request by MatthewKennedy**
- [x] Setup migrations to account for anyone already using the `spree_digital`.
- [x] Move model files over unchanged.
- [x] Move logic that was added to existing classes by a decorator into the Spree core existing models.
- [x] Migrate Preferences to core moved to store JSON in the basic working form.
- [x] Setup Storefront API to access authorisation of a download link.
- [x] Setup Platform API - CRUD for a digital asset.
- [x] Figure out the use case of the preferences and if they need to be any more complex than the store the item was sold in.
- [x] Add update platform API docs.
- [x] Migrate relevant parts of the spec test suite
- [x] Write a sample bit of javascript to test out the file download link from the actual browser side, test Safari, Chrome, Fire Fox for file types.
- [x] Check for any TODO's left in new code

In this PR we venture into using Rails Active store on a JSON or JSONb column for use in store wide settings, this raises the issue that settings need to be of the following types: `integer`, `boolean`, `string`, `array` through ActiveRecord::BaseStore we would only get strings and thus this PR brings in a third party gem [ActiveRecord::TypedStore](https://github.com/byroot/activerecord-typedstore), this needs to be considered before going down this path.

### NOTE:
I have kept `access_counter` in the code I was thinking, depending how the download link is sent, either in the completed order email, it could be useful to stop DDoS attacks where someone could set up a script to hammer a download link until the servers crap out, or share the link publicly and essentially render your product useless.

But also if you sell stock photo services where you need to login and download, you should also be able to set null value for the downloads limit so that it can be downloaded any number of times by the logged in user.

### Changes To `platform_context.rb` For Auto Generated API Docs

To document the uploading of digital goods I had to make some changes to the  `platform_context.rb` allowing the passing of a different request body content-type, the default is the standard application/json but for file uploads it really needs a form-data type, and thus now there is a way to override this and pass the new type, this should come in handy for documenting asset uploads.

The docs generate import into postman set up to upload a file easy, and also show upload button in Stoplight.io

### CODE REVIEW

File: core/app/models/spree/digital_link.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
+module Spree
+  class DigitalLink < ActiveRecord::Base
```
Comment (damianlegawiec) [inline core/app/models/spree/digital_link.rb:2]:
```suggestion
  class DigitalLink < Spree::Base
```

File: core/app/models/spree/digital.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+module Spree
+  class Digital < ActiveRecord::Base
```
Comment (damianlegawiec) [inline core/app/models/spree/digital.rb:2]:
```suggestion
  class Digital < Spree::Base
```

File: core/app/models/spree/variant.rb [hunk 3]
```diff
@@ -302,8 +304,22 @@ def backordered?
...
+      digitals.present?
+    end
+
     private
 
+    # :dependent => :destroy needs to be handled manually
+    # spree does not delete variants, just marks them as deleted?
+    # optionally keep digitals around for customers who require continued access to their purchases
+    def destroy_digital
+      return if Spree::DigitalConfiguration[:keep_digitals]
```
Comment (damianlegawiec) [inline core/app/models/spree/variant.rb:318]: @MatthewKennedy let's drop the `DigitalConfiguration` and move that preference to `Store` model

File: core/app/models/spree/variant.rb [hunk 4]
```diff
@@ -302,8 +304,22 @@ def backordered?
...
+
+      digitals.map(&:destroy)
```
Comment (damianlegawiec) [inline core/app/models/spree/variant.rb:320]:
```suggestion
      digitals.destroy_all
```

File: core/app/models/spree/app_configuration.rb [hunk 5]
```diff
@@ -81,6 +81,12 @@ def searcher_class
     preference :storefront_taxons_path, :string, default: 't'
     preference :storefront_pages_path, :string, default: 'pages'
 
+    # Spree Digital Asset Configurations
+    preference :authorized_clicks, :integer, default: 3 # number of times a customer can download a digital file
```
Comment (damianlegawiec) [inline core/app/models/spree/app_configuration.rb:85]: @MatthewKennedy this should be set on item level

File: core/app/models/spree/app_configuration.rb [hunk 6]
```diff
@@ -81,6 +81,12 @@ def searcher_class
...
+    preference :authorized_days, :integer, default: 2 # number of days after initial purchase the customer can download a file
```
Comment (damianlegawiec) [inline core/app/models/spree/app_configuration.rb:86]: same here

File: core/app/models/spree/app_configuration.rb [hunk 7]
```diff
@@ -81,6 +81,12 @@ def searcher_class
...
+    preference :keep_digitals, :boolean, default: false # should digitals be kept around after the associated product is destroyed
```
Comment (damianlegawiec) [inline core/app/models/spree/app_configuration.rb:87]: @MatthewKennedy we should always keep digitals

File: core/app/models/spree/app_configuration.rb [hunk 8]
```diff
@@ -81,6 +81,12 @@ def searcher_class
...
+    preference :s3_expiration_seconds, :integer, default: 10 # number of seconds before an s3 link expires
```
Comment (damianlegawiec) [inline core/app/models/spree/app_configuration.rb:88]: @MatthewKennedy this can be removed, not needed

Comment (MatthewKennedy) [reply core/app/models/spree/app_configuration.rb:88]: I'll get to this!

File: core/app/models/spree/variant.rb (hunk 4 above)
Comment (MatthewKennedy) [reply core/app/models/spree/variant.rb:320]: @damianlegawiec If "we should always keep digitals" this whole method is not required, would you agree?

File: core/app/models/spree/digital_link.rb [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    def reset!
+      update_column :access_counter, 0
+      update_column :created_at, Time.now
+    end
+
+    private
+
+    # Populating the secret automatically and zero'ing the access_counter (otherwise it might turn out to be NULL)
+    def set_defaults
+      self.secret = SecureRandom.hex(15)
```
Comment (damianlegawiec) [inline core/app/models/spree/digital_link.rb:40]: @MatthewKennedy let's use https://api.rubyonrails.org/classes/ActiveRecord/SecureToken/ClassMethods.html

**Comment by MatthewKennedy** [comment]
@damianlegawiec Should we rename the `secret` column to `token` to be in-keeping with the use of token in other parts of Spree such as in the Wishlists.

Additionally, what do you think of naming the new active store preferences json column `configs` or `configurations` this way we are never conflicting with any existing `preferences` column that might of been used in the past for model preferences, such as is used on the spree_payment_methods table.

**Comment by damianlegawiec** [comment]
> @damianlegawiec Should we rename the `secret` column to `token` to be in-keeping with the use of token in other parts of Spree such as in the Wishlists.

Yes, let's do this, please


> Additionally, what do you think of naming the new active store preferences json column `configs` or `configurations` this way we are never conflicting with any existing `preferences` column that has been used in the past for model preferences, such as is used on the spree_payment_methods table.

Maybe let's use `settings`?

File: core/app/models/spree/order.rb [hunk 10]
```diff
@@ -663,6 +663,27 @@ def has_free_shipping?
               spree_promotion_actions: { type: 'Spree::Promotion::Actions::FreeShipping' }).exists?
     end
 
+    # all products are digital
```
Comment (damianlegawiec) [inline core/app/models/spree/order.rb:666]: @MatthewKennedy can we move this into a separate module called `Spree::Order::Digital` eg. https://github.com/spree/spree/blob/main/core/app/models/spree/order/store_credit.rb

Order model is quite big already 😂

File: core/app/models/spree/calculator/shipping/digital_delivery.rb [hunk 11]
```diff
@@ -0,0 +1,22 @@
+require_dependency 'spree/shipping_calculator'
+
+module Spree
```
Comment (damianlegawiec) [inline core/app/models/spree/calculator/shipping/digital_delivery.rb:3]: @MatthewKennedy let's also add `Digital` category here:

https://github.com/spree/spree/blob/main/core/app/services/spree/seeds/shipping_categories.rb

```ruby
Spree::ShippingCategory.find_or_create_by!(name: 'Digital')
```

**Review by damianlegawiec — APPROVED** [review]
🚀

**Pull request merged**